### PR TITLE
Java: make a separate threat model kind for reverse DNS sources

### PR DESCRIPTION
**Pull request by owen-mc**
I'm new to threat models, so I may be missing something that is required. In particular, is it okay to have a threat model kind that isn't part of the hierarchy controlled by [this file](https://github.com/github/codeql/blob/3d037e73f0d76b32ce9bc508246a001588fdd20f/shared/threat-models/ext/threat-model-grouping.model.yml)?

Also, I searched for other reverse DNS sources and only found [this one in C#](https://github.com/github/codeql/blob/e2a47e7c1872e0593022d7c09b581a91b6d44530/csharp/ql/lib/semmle/code/csharp/security/dataflow/ConditionalBypassQuery.qll#L74-L79), which seems to be only be used in one query currently. Any opinion on whether that should be made a source that applies to all queries and added to this new threat model kind? (I think this should be follow-up work.)

### CODE REVIEW

File: java/ql/lib/change-notes/2024-06-14-reverse-dns-separate-threat-model-kind.md
Comment (aeisenberg) [inline java/ql/lib/change-notes/2024-06-14-reverse-dns-separate-threat-model-kind.md:4]: Might make sense to include an example of how to use this new threat model.

File: shared/threat-models/ext/threat-model-grouping.model.yml
Comment (aeisenberg) [inline shared/threat-models/ext/threat-model-grouping.model.yml:26]: This is probably fine the way it is, though it could be confusing if we start having lots of un-rooted threat models. 

The reason this works is that we want the ability for people to define their own threat models (and even threat model groupings) that are neither `local` nor `remote`.

Not for this PR, but I wonder if we should refactor so that there is an `all` threat model that can be used to include all threat models defined in the standard libraries (though I'm not sure `all` may already be special threat model that is hard coded, so we could bike shed a different name).

**Review by aeisenberg — COMMENTED** [review]
Looks generally good (with one suggestion). I looked at the threat model part of this. Someone more familiar with the ql should review the rest.

File: shared/threat-models/ext/threat-model-grouping.model.yml
Comment (michaelnebel) [reply shared/threat-models/ext/threat-model-grouping.model.yml:26]: There already exist a hardcoded `all` threat model that enables "everything" also threat models that are not listed in any groups.

Comment (michaelnebel) [reply shared/threat-models/ext/threat-model-grouping.model.yml:26]: I think we should strive for documenting all the threat models that we supply as a part of our external documentation. It is also a good idea to list them in the shared file as is done in this PR.

File: java/ql/lib/change-notes/2024-06-14-reverse-dns-separate-threat-model-kind.md
Comment (owen-mc) [reply java/ql/lib/change-notes/2024-06-14-reverse-dns-separate-threat-model-kind.md:4]: I imagine it depends on how you are running CodeQL (code-scanning, CLI, extension), so it might be best to link to some docs for this. (Do we put links in change notes? And do we have docs on this?)

File: docs/codeql/reusables/threat-model-description.rst
Comment (aeisenberg) [inline docs/codeql/reusables/threat-model-description.rst:14]: I'm assuming this is kotlin only (though maybe Java as well?).

```suggestion
- ``android`` which represents reads from external files in Android (``android-external-storage-dir``) and parameter of an entry-point method declared in a ``ContentProvider`` class (``contentprovider``). Currently only used by Kotlin.
```

File: docs/codeql/reusables/threat-model-description.rst
Comment (aeisenberg) [inline docs/codeql/reusables/threat-model-description.rst:17]: Minor copy-editing:

```suggestion
- ``database-access-result`` which represents a database access. Currently only used by JavaScript.
- ``file-write`` which represents opening a file in write mode. Currently only used in C#.
- ``reverse-dns`` which represents reverse DNS lookups. Currently only used in Java.
```

**Review by aeisenberg — DISMISSED** [review]
LGTM. Thanks for updating the documentation.

File: docs/codeql/reusables/threat-model-description.rst
Comment (owen-mc) [reply docs/codeql/reusables/threat-model-description.rst:14]: It's kind of hard to tell, but at least one of the tests for this is written in java. I'll put "Java/Kotlin".

**Comment by owen-mc** [comment]
@aeisenberg Thanks. Note that `windows-registry` is currently only used by C#. Any suggestion for how to shoe-horn that info in? Also, I might well make a `reverse-dns` source for C# after this PR is merged - I assume that I should then remove the " Currently only used in Java." part, rather than making it "C# and Java"?

File: docs/codeql/reusables/threat-model-description.rst
Comment (aeisenberg) [inline docs/codeql/reusables/threat-model-description.rst:8]: Not great, but I think this expresses the information you're looking for.

```suggestion
- ``local`` which represents data from local files (``file``), command-line arguments (``commandargs``), database reads (``database``), environment variables(``environment``) and Windows registry values ("windows-registry"). Windows registry values are used by C# only.
```

**Review by aeisenberg — APPROVED** [review]

**Comment by owen-mc** [comment]
@aeisenberg Note that I've reverted one line that I had changed in the docs - it no longer explains about the subcategories `response` and `request` of `remote`, since no languages use those categories yet.

**Review by aschackmull — APPROVED** [review]
QL LGTM

**Pull request merged**
